rust/parquet/src/arrow/levels.rs

```diff
@@ -294,40 +242,1158 @@ impl LevelInfo {
+            repetition: None,
+            array_offsets: (0..=5).collect(),
+            array_mask: vec![true, true, true, true, true],
+            max_definition: 0,
+            is_list: false,
+            is_nullable: false,
+        };
+        let array_offsets: Vec<i64> = (0..=5).collect();
+        let array_mask = vec![true, false, true, true, false];
+
+        let levels = parent_levels.calculate_child_levels(
+            array_offsets.clone(),
+            array_mask.clone(),
+            false,
+            true,
+        );
+        let expected_levels = LevelInfo {
+            definition: vec![1, 0, 1, 1, 0],
+            repetition: None,
+            array_offsets,
+            array_mask,
+            max_definition: 1,
+            is_list: false,
+            is_nullable: true,
+        };
+        assert_eq!(&levels, &expected_levels);
+    }
+
+    #[test]
+    fn test_calculate_array_levels_1() {
+        // if all array values are defined (e.g. batch<list<_>>)
+        // [[0], [1], [2], [3], [4]]
+        let parent_levels = LevelInfo {
+            definition: vec![0; 5],
+            repetition: None,
+            array_offsets: vec![0, 1, 2, 3, 4, 5],
+            array_mask: vec![true, true, true, true, true],
+            max_definition: 0,
+            is_list: false,
+            is_nullable: false,
+        };
+        let array_offsets = vec![0, 2, 2, 4, 8, 11];
+        let array_mask = vec![true, false, true, true, true];
+
+        let levels = parent_levels.calculate_child_levels(
+            array_offsets.clone(),
+            array_mask,
+            true,
+            true,
+        );
+        // array: [[0, 0], _1_, [2, 2], [3, 3, 3, 3], [4, 4, 4]]
+        // all values are defined as we do not have nulls on the root (batch)
+        // repetition:
+        //   0: 0, 1
+        //   1:
+        //   2: 0, 1
+        //   3: 0, 1, 1, 1
+        //   4: 0, 1, 1
+        let expected_levels = LevelInfo {
+            definition: vec![2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2],
+            repetition: Some(vec![0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1]),
+            array_offsets,
+            array_mask: vec![
+                true, true, false, true, true, true, true, true, true, true, true, true,
+            ],
+            max_definition: 2,
+            is_list: true,
+            is_nullable: true,
+        };
+        assert_eq!(&levels.definition, &expected_levels.definition);
+        assert_eq!(&levels.repetition, &expected_levels.repetition);
+        assert_eq!(&levels.array_offsets, &expected_levels.array_offsets);
+        assert_eq!(&levels.max_definition, &expected_levels.max_definition);
+        assert_eq!(&levels.is_list, &expected_levels.is_list);
+        assert_eq!(&levels.is_nullable, &expected_levels.is_nullable);
+        assert_eq!(&levels, &expected_levels);
+    }
+
+    #[test]
+    fn test_calculate_array_levels_2() {
+        // If some values are null
```

Comment:
Some of these comments need updating. They're from 2-3 iterations ago, and though correct, they might be referring to incorrect definition levels relative to the test cases.